actionpack/lib/action_controller/renderer.rb

```diff
@@ -0,0 +1,66 @@
+module ActionController
+  class Renderer
+    class_attribute :controller, :defaults
+    attr_reader :env
+
+    class << self
+      delegate :render, to: :new
+
+      def for(controller)
+        Class.new self do
+          self.controller = controller
+          self.defaults = {
+            http_host: 'example.org',
+            https: false,
+            method: 'get',
+            script_name: '',
+            'rack.input' => ''
+          }
+        end
+      end
+    end
+
+    def initialize(env = {})
+      @env = normalize_keys(defaults).merge normalize_keys(env)
```

Comment:
Why do we need to normalize keys of environment? To match expected format of rack env by the rest.

Or why do we need defaults? It's the minimal environment in which different helpers and request would work without exceptions each time.